lib/charms/sdcore_nrf/v0/fiveg_nrf.py

```diff
@@ -0,0 +1,296 @@
+
+
+class NRFRequires(Object):
+    """Class to be instantiated by the NRF requirer charm."""
+
+    on = NRFRequirerCharmEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handler triggered on relation changed event.
+
+        Args:
+            event (RelationChangedEvent): Juju event.
+
+        Returns:
+            None
+        """
+        if remote_app_relation_data := self._get_remote_app_relation_data(event.relation):
+            self.on.nrf_available.emit(url=remote_app_relation_data["url"])
+
+    @property
+    def nrf_url(self) -> Optional[str]:
+        """Returns NRF url.
+
+        Returns:
+            str: NRF url.
+        """
+        if remote_app_relation_data := self._get_remote_app_relation_data():
+            return remote_app_relation_data.get("url")
+        return None
+
+    def _get_remote_app_relation_data(self, relation: Optional[Relation] = None) -> Optional[dict]:
+        """Get relation data for the remote application.
+
+        Args:
+            Relation: Juju relation object (optional).
+
+        Returns:
+            dict: Relation data for the remote application.
+            or None if the relation data is invalid.
+        """
+        relation = relation or self.model.get_relation(self.relation_name)
+        if not relation:
+            logger.warning(f"No relation: {self.relation_name}")
+            return None
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relation_name)
+            return None
+        remote_app_relation_data = dict(relation.data[relation.app])
+        if not data_is_valid(remote_app_relation_data):
+            logger.error("Invalid relation data: %s", remote_app_relation_data)
+            return None
+        return remote_app_relation_data
+
+
+class NRFProvides(Object):
+    """Class to be instantiated by the charm providing the NRF data."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Init."""
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+        self.charm = charm
+
+    def set_nrf_information(
+        self,
+        url: str,
+        event: Optional[RelationJoinedEvent] = None,
+        update_all_relations: bool = False,
+    ) -> None:
+        """Sets NRF url in the application(s) relation data.
+
+        Args:
+            url (str): NRF url.
+            event (Optional[RelationJoinedEvent]): Juju event.
+            update_all_relations (bool): Whether to update all relations or not.
+
+        Returns:
+            None
+        """
```

Comment:
I think I would prefer to see at least 2 methods for this. Having a flag like `update_all_relations` to change the behavior of the method is a bit weird, and I think it would be clearer to have a separate method `set_nrf_information_in_all_relations`.

WDYT?

Reply:
Agreed, also for IDE completion will be better.